### PR TITLE
Give actions descriptions

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Adds a new model called `DescribedItem`, which contains the description-object relationship across any kind of instance. This work is very closely tied to the rationale behind https://github.com/PostHog/posthog/pull/8331. See discussion there for context behind decision to refactor enterprise features out at a property level instead of the model level.

A bunch of older models still use inline `description` fields, but for the purposes of bringing the enterprise action model up to date with its event definition/property definition counterpart, this PR only adds descriptions to the action model.

Future work will tackle porting over the rest of the below models to use this new global relationship table. 
- `models/cohort.py`                 
- `models/dashboard.py`              
- `models/insight.py`                
- `ee/models/event_definition.py`    
- `ee/models/property_definition.py`  

The new workflow for creating/editing an action

https://user-images.githubusercontent.com/13460330/151042268-8c283ab3-2aff-4bef-a0ac-ba8378649b15.mov

## How did you test this code?

Add new model tests and update existing action tests.

### CODE REVIEW

File: posthog/models/action.py [hunk 1]
```diff
@@ -23,6 +23,7 @@ class Meta:
 
     name: models.CharField = models.CharField(max_length=400, null=True, blank=True)
     team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
+    description: models.CharField = models.CharField(max_length=400, blank=True)
```
Comment (rcmarron) [inline posthog/models/action.py:26]: Right now, we cap the limit to 400 characters, but if the user types over 400 characters, we fail silently on the front-end.

I don't have a strong preference, but I think we should either handle the error + make it clear to the user, or if we don't really want to enforce a max_length, then use `models.TextField` .

**Review by rcmarron — DISMISSED** [review]
Looks good - found one bug with description length below.

Also, this may be a question for @paolodamico, but I wonder if this should be an enterprise feature (and therefore the field would live on a new `EnterpriseAction` model). That way it would mirror the license needed for the event description. Thoughts?

File: posthog/models/action.py (hunk 1 above)
Comment (Twixes) [reply posthog/models/action.py:26]: `models.TextField` is generally what we use for descriptions I think.
Regarding the frontend, at least length validation is simple enough to just slap `maxLength` on the input component.

**Comment by Twixes** [comment]
Scope-creeping observation: the name field was already clunky, but this description box is even more so. How about reusing the name/description style from the insight page (utilizing `EditableField`)?

File: posthog/models/action.py (hunk 1 above)
Comment (alexkim205) [reply posthog/models/action.py:26]: 👍 to slapping `maxLength`

**Comment by alexkim205** [comment]
@rcmarron re EnterpriseAction: I thought about doing it this way, but I don't think it's too much of a tradeoff to continue working within the existing model. I think I prefer adding a field here for now rather than adding an entirely new model bc there is no difference to the end user how we choose to organize our data and any feature access is already handled on the frontend and and it lets me move a little quicker.

----

New Look:

https://user-images.githubusercontent.com/13460330/151086648-9f08f6de-e1e7-4507-8a7d-cf56c7185bb6.mov

**Comment by rcmarron** [comment]
> @rcmarron re EnterpriseAction: I thought about doing it this way, but I don't think it's too much of a tradeoff to continue working within the existing model. I think I prefer adding a field here for now rather than adding an entirely new model bc there is no difference to the end user how we choose to organize our data and any feature access is already handled on the frontend and and it lets me move a little quicker.

Got it, I think the reason that we have the models split is so that the code for the enterprise features can be in the `ee/` folder (which has the non-open-source license). So if we're planning for it to be an enterprise feature, then I think we would want to keep it in that folder.

**Comment by paolodamico** [comment]
Makes sense that taxonomy on actions be a paid feature too

**Comment by alexkim205** [comment]
Yup makes sense, I'm going to bring this into draft again, and put together a new EnterpriseAction model. Still trying to understand what Django takes care of under the hood and what it doesn't so pls bear w/ my slower iterations 😓

**Comment by Twixes** [comment]
Just making sure – is this currently awaiting a rework before review? (BTW new editable header looks great)

**Comment by alexkim205** [comment]
> Just making sure – is this currently awaiting a rework before review? (BTW new editable header looks great)

Thanks! and yup I'm going to revisit this after https://github.com/PostHog/posthog/pull/8331 is merged in so that we can set a better standard for splitting out ee code. I think descriptions here would follow a similar pattern. I'll keep ya posted

**Comment by alexkim205** [comment]
Update @Twixes , going to pare this PR down into just adding a description to action to keep this PR small and non blocking. 

Edit: just put out a new rev!

File: posthog/models/described_item.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """
+
+    description: models.TextField = models.TextField(blank=True)
+    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
+
+    content_type: models.ForeignKey = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    # Primary key value of related model. Query by this to get all tags for specific model. This is a charfield because
+    # there we don't have a standard way of storing objects. Some models use positive integer ids and others use UUID's.
+    object_id: models.CharField = models.CharField(max_length=400)
+    content_object: GenericForeignKey = GenericForeignKey("content_type", "object_id")
```
Comment (Twixes) [inline posthog/models/described_item.py:42]: This looks like a LOT of complexity for the single meaningful field in this model: `description`. I think it makes sense to separate significant pieces of functionality into EE, like advanced event type metadata, but just `description` isn't really significant, making this kind of overkill… I just don't see benefits in this very specific plan

**Review by Twixes — DISMISSED** [review]
The UI changes look good but I'm not sure if there are benefits to the Django models strategy that make the added complexity worth it.

File: posthog/models/described_item.py (hunk 2 above)
Comment (alexkim205) [reply posthog/models/described_item.py:42]: Makes sense! Thanks for clarifying at standup. I'll revert this to how it used to be

**Review by mariusandra — DISMISSED** [review]
Still some tests look broken... and I could save a new action with no title. Might we want to enforce something here?

However, I can just have one action with the undefined title `null`. The second one gets an error :D 



UX-wise, can we make the name textfield default to looking like a field for new actions or when it's empty? It was super easy to just overlook it in the gif above and go directly to a field that looked interactive.

**Comment by alexkim205** [comment]
Thanks for the lookaround, and good catch. The new flow shouldn't let you get anywhere near saving if the name is empty.


https://user-images.githubusercontent.com/13460330/152075869-a2136b84-3f84-4035-9619-cb6fd0af12b8.mov

File: frontend/src/scenes/actions/ActionEdit.tsx [hunk 3]
```diff
@@ -61,27 +65,49 @@ export function ActionEdit({ action: loadedAction, id, onSave, temporaryToken }:
...
+                        onSave={(description) => {
+                            setAction({ ...action, description })
+                            setEdited(!!description)
+                        }}
+                        data-attr="action-description"
+                        compactButtons
+                        maxLength={600} // No limit on backend model, but enforce shortish description
+                        isGated={!hasAvailableFeature(AvailableFeature.INGESTION_TAXONOMY)}
+                        saveButtonText="Set"
+                    />
```
Comment (Twixes) [inline frontend/src/scenes/actions/ActionEdit.tsx:100]: It's a bit confusing that the name field here is persistently editable when creating the action, but the description field isn't. Would be more intuitive if they behaved the same.

**Review by Twixes — CHANGES_REQUESTED** [review]
Just two notes

**Review by Twixes — APPROVED** [review]
Implemented the few changes to get this out the door

**Pull request merged**